File: SMACrossoverStrategy.java
import java.util.List;

/**
 * SMA Crossover strategy — generates buy/sell signals based on the crossover of
 * a short-period Simple Moving Average over a long-period SMA.
 *
 * <ul>
 *   <li><b>Golden Cross (BUY):</b> short SMA crosses above long SMA</li>
 *   <li><b>Death Cross (SELL):</b> short SMA crosses below long SMA</li>
 * </ul>
 *
 * <p>Common configurations: 50/200 (long-term trend), 20/50 (medium-term swing).</p>
 */
public final class SMACrossoverStrategy implements TradingStrategy {

    private final int shortPeriod;
    private final int longPeriod;

    public SMACrossoverStrategy(int shortPeriod, int longPeriod) {
        if (shortPeriod <= 0 || longPeriod <= 0) {
            throw new IllegalArgumentException("Periods must be positive: short=%d, long=%d"
                    .formatted(shortPeriod, longPeriod));
        }
        if (shortPeriod >= longPeriod) {
            throw new IllegalArgumentException("Short period (%d) must be less than long period (%d)"
                    .formatted(shortPeriod, longPeriod));
        }
        this.shortPeriod = shortPeriod;
        this.longPeriod = longPeriod;
    }

    @Override
    public String getName() {
        return "SMA Crossover (%d/%d)".formatted(shortPeriod, longPeriod);
    }

    @Override
    public Signal evaluate(List<PriceData> data, int currentIndex) {
        validateEvaluationInputs(data, currentIndex);

        if (!isWarmedUp(currentIndex)) {
            return Signal.HOLD;
        }

        var currentShortSMA = TechnicalIndicators.sma(data, currentIndex, shortPeriod);
        var currentLongSMA = TechnicalIndicators.sma(data, currentIndex, longPeriod);
        var prevShortSMA = TechnicalIndicators.sma(data, currentIndex - 1, shortPeriod);
        var prevLongSMA = TechnicalIndicators.sma(data, currentIndex - 1, longPeriod);

        boolean shortAboveNow = currentShortSMA.compareTo(currentLongSMA) > 0;
        boolean shortAbovePrev = prevShortSMA.compareTo(prevLongSMA) > 0;

        if (shortAboveNow && !shortAbovePrev) return Signal.BUY;
        if (!shortAboveNow && shortAbovePrev) return Signal.SELL;

        return Signal.HOLD;
    }

    @Override
    public int getWarmupPeriod() {
        return longPeriod;
    }

    public int getShortPeriod() { return shortPeriod; }
    public int getLongPeriod()  { return longPeriod; }

    @Override
    public String toString() {
        return getName();
    }
}

File: ios/StockAnalyzer/StockAnalyzer/Strategies/SMACrossoverStrategy.swift
import Foundation

struct SMACrossover50_200Strategy: TradingStrategy {
    let name = "SMA Crossover (50/200)"
    let description = "Golden Cross / Death Cross: Buys when 50-day SMA crosses above 200-day SMA"
    let warmupPeriod = 200

    func evaluate(data: [PriceData], index: Int, indicators: StrategyIndicators) -> TradeAction {
        guard index >= 1,
              let sma50 = indicators.sma50[index],
              let sma200 = indicators.sma200[index],
              let prevSma50 = indicators.sma50[index - 1],
              let prevSma200 = indicators.sma200[index - 1] else {
            return .hold
        }

        if prevSma50 <= prevSma200 && sma50 > sma200 {
            return .buy
        }

        if prevSma50 >= prevSma200 && sma50 < sma200 {
            return .sell
        }

        return .hold
    }
}

struct SMACrossover20_50Strategy: TradingStrategy {
    let name = "SMA Crossover (20/50)"
    let description = "Medium-term trend following: Buys when 20-day SMA crosses above 50-day SMA"
    let warmupPeriod = 50

    func evaluate(data: [PriceData], index: Int, indicators: StrategyIndicators) -> TradeAction {
        guard index >= 1,
              let sma20 = indicators.sma20[index],
              let sma50 = indicators.sma50[index],
              let prevSma20 = indicators.sma20[index - 1],
              let prevSma50 = indicators.sma50[index - 1] else {
            return .hold
        }

        if prevSma20 <= prevSma50 && sma20 > sma50 {
            return .buy
        }

        if prevSma20 >= prevSma50 && sma20 < sma50 {
            return .sell
        }

        return .hold
    }
}
